### PR TITLE
[EH Perf improvement] Translate Reactor's no-backpressure to prefetch in amqp Flow

### PR DESCRIPTION
**Pull request by anuchandy**
When there is no back-pressure (as a result of using a reactor-operator designed for no back-pressure) we can set credit to `prefetch`. By reactive design, once the downstream operator sends a request with `MAX_VALUE`, future requests (if any) will be `MAX_VALUE` as well, so in this code-path, credit is set to the max value capped by prefetch.

Below is an example for an application using no-backpressure opeator [take(int)](https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Flux.html#take-long-):

```java
public class NoBackPressure {
    public static void run(String connectionString,
                           String eventHubName,
                           String consumerGroup,
                           String partitionId,
                           int preFetch,
                           int count) {
        final EventHubConsumerAsyncClient receiver = new EventHubClientBuilder()
            .connectionString(connectionString, eventHubName)
            .transportType(AmqpTransportType.AMQP)
            .prefetchCount(preFetch)
            .consumerGroup(consumerGroup)
            .buildAsyncConsumerClient();

        long startNanoTime = System.nanoTime();
        int [] events = new int[1];
        receiver.receiveFromPartition(partitionId, EventPosition.earliest())
            .take(count)
            .doOnNext(e -> events[0]++)
            .blockLast();
        long endNanoTime = System.nanoTime();
        System.out.println("Took " + Duration.ofNanos(endNanoTime - startNanoTime).getSeconds() + " seconds to receive " + events[0] + " events");
        System.out.println("Closing-Client");
        receiver.close();
        System.out.println("Closed-Client");
    }
}
```

### CODE REVIEW

**Review by conniey — APPROVED** [review]

**Comment by YijunXieMS** [comment]
Do we have any code branch that has backpressure (the request size is not MAX_VALUE)?. We may need to increment link credits by the prefetch that users pass in with default value 500, instead of the default prefetch of some Reactor APIs, which is usually 32 by default and subsequent request size 24. This might lead to poor performance.

**Comment by anuchandy** [comment]
@YijunXieMS  thank you for taking a look. When there is "No Back-Pressure", i.e. reactor requests MAX_VALUE, the value we use for credit is the prefetch value that user-provided, provided via `EventHubClientBuilder::prefetchCount(int)`.

Yes, there is a code path with "Back-Pressure", i.e. reactor requests a non-MAX_VALUE, say X, in that case, the credit is set to  
`count(queue-elements-tobe-delivered) - X`, but we can potentially optimize this path to reduce network traffic, I'll send another pr for that code path after some thoughts.

**Review by YijunXieMS — APPROVED** [review]

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/AmqpReceiveLinkProcessor.java [hunk 1]
```diff
@@ -584,7 +584,7 @@ private int getCreditsToAdd() {
         if (subscriber == null || request == 0) {
             credits = 0;
         } else if (request == Long.MAX_VALUE) {
-            credits = 1;
+            credits = prefetch;
```
Comment (YijunXieMS) [inline sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/AmqpReceiveLinkProcessor.java:587]: I guess the value of `prefetch` is the value of `preFetch` in your demo code?

Comment (anuchandy) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/AmqpReceiveLinkProcessor.java:587]: that's correct.

Comment (srnagar) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/AmqpReceiveLinkProcessor.java:587]: The JavaDoc for this method should be updated to reflect this behavior. Apparently, this behavior of adding just 1 credit at a time is similar to what Track 1 did. There may have been a reason for doing it this way that @conniey may be aware of.

Comment (srnagar) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/AmqpReceiveLinkProcessor.java:587]: Also, have we measured the perf improvement with this change and has this made the client less chatty?

Comment (YijunXieMS) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/AmqpReceiveLinkProcessor.java:587]: Another question is, what's the theoretical max number of messages in the cache when we add credits in hundreds.
For instance, if the cache already has a lot of messages and we add a few hundred credits after each message is consumed?

Comment (anuchandy) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/AmqpReceiveLinkProcessor.java:587]: > srnagar 
Also, have we measured the perf improvement with this change and has this made the client less chatty?

@srnagar, yep, I've used the `receiveevents` test in the perf framework which exercises the no-backpressure code path. The performance of T2 in this code-path improved drastically and T1 and T2 are very close now. 

Below is the result of perf testing this specific code path with a prefetch of 100 and a count of 1800 (the test is the same as `receiveevents` perf test, I slightly modified it to be more debuggable and print extra info).

[Executed for 5 times]
 
```
TRACK_1: Start--
Took 14 seconds to receive 1800 events
TRACK_1: End--

TRACK_2: Start--
Took 9 seconds to receive 1800 events
TRACK_2: End--

TRACK_1: Start--
Took 5 seconds to receive 1800 events
TRACK_1: End--

TRACK_2: Start--
Took 5 seconds to receive 1800 events
TRACK_2: End--

TRACK_1: Start--
Took 5 seconds to receive 1800 events
TRACK_1: End--

TRACK_2: Start--
Took 5 seconds to receive 1800 events
TRACK_2: End--

TRACK_1: Start--
Took 5 seconds to receive 1800 events
TRACK_1: End--

TRACK_2: Start--
Took 12 seconds to receive 1800 events
TRACK_2: End--

TRACK_1: Start--
Took 8 seconds to receive 1800 events
TRACK_1: End--

TRACK_2: Start--
Took 7 seconds to receive 1800 events
TRACK_2: End--
```

Comment (anuchandy) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/AmqpReceiveLinkProcessor.java:587]: > srnagar .... Apparently, this behavior of adding just 1 credit at a time is similar to what Track 1 did. There may have been a reason for doing it this way that @conniey may be aware of.

Connie and I visited that track-1 code, track-1 has a special code path that adds 1 credit at a time, but that only accumulates the credits locally. Once the accumulated credits hit a cap, the actual Flow will be sent. So, track2 won't send one flow for each credit.

Comment (anuchandy) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/AmqpReceiveLinkProcessor.java:587]: > YijunXieMS 
Another question is, what's the theoretical max number of messages in the cache when we add credits in hundreds.
For instance, if the cache already has a lot of messages and we add a few hundred credits after each message is consumed?

`what's the theoretical max number of messages in the cache` - It depends on the operators that the user applied in the chain in their application side, different operators can have their own queue, holding messages to be delivered for their immediate subscriber. Our processor also has its own queue (say q1), when we receive a "request" for messages from the downstream, the credit will be computed using the size of q1 and "request"ed amount, that's the back-pressure aware flow, MAX_VALUE is the no-back-pressure flow.

**Pull request merged**